File: go-mongo/models/product.go
package models

type ProductLanguage struct {
	Type         string       `json:"@type,omitempty" validate:"required" bson:"@type,omitempty"`
	ID           string       `json:"id" validate:"required" bson:"id" binding:"required"`
	Href         string       `json:"href,omitempty" bson:"href,omitempty"`
	LanguageCode string       `json:"languageCode" bson:"languageCode" binding:"required"`
	Name         string       `json:"name,omitempty" bson:"name,omitempty"`
	Version      string       `json:"version,omitempty" bson:"version,omitempty"`
	LastUpdate   string       `json:"lastUpdate,omitempty" bson:"lastUpdate,omitempty"`
	Attachment   []Attachment `json:"attachment,omitempty" bson:"attachment,omitempty"`
}

type Attachment struct {
	Type        string      `json:"@type,omitempty" validate:"required" bson:"@type,omitempty"`
	ID          string      `json:"id,omitempty" validate:"required" bson:"id,omitempty" binding:"required"`
	Href        string      `json:"href,omitempty" bson:"href,omitempty"`
	Description string      `json:"description,omitempty" bson:"description,omitempty"`
	MimeType    string      `json:"mimeType,omitempty" bson:"mimeType,omitempty"`
	Name        string      `json:"name,omitempty" bson:"name,omitempty"`
	Url         string      `json:"url,omitempty" bson:"url,omitempty"`
	RedirectUrl string      `json:"redirectUrl,omitempty" bson:"redirectUrl,omitempty"`
	DisplayInfo DisplayInfo `json:"displayInfo,omitempty" bson:"displayInfo,omitempty"`
}

type DisplayInfo struct {
	ValueType string   `json:"valueType,omitempty" bson:"valueType,omitempty"`
	Value     []string `json:"value,omitempty" bson:"value,omitempty"`
}

File: src/model/product.ts
export interface IProduct {
  id?: string
  name: string
  '@type'?: string
  href?: string
  status?: string
  description?: string
  price?: IPrice
  supportLanguage?: IProductLanguage[]
  startDate?: Date
  endDate?: Date
  deleteDate?: Date | null
}

export interface IProductResponse {
  id?: string
  name?: string
  '@type'?: string
  href?: string
  status?: string
  description?: string
  price?: IPrice
  startDate?: Date
  endDate?: Date
  deleteDate?: Date | null
  supportLanguage?: ProductLanguage[]
}

export interface IProductLanguage {
  '@referentType'?: string
  id?: string
  languageCode?: string
}

export interface IPrice {
  price?: number
  priceType?: string
  unit?: string
  tax?: number
  taxType?: string
  discount?: number
  discountType?: string
  fullPrice?: number
  fullPriceType?: string
  fullPriceEndDate?: Date
  fullPriceStartDate?: Date
  fullPriceActive?: boolean
}

export enum ProductStatus {
  ACTIVE = 'active',
  INACTIVE = 'inActive',
}

export interface ProductLanguage {
  '@type'?: string
  id: string
  href?: string
  languageCode: string
  name?: string
  version?: string
  lastUpdate?: string
  attachment?: Attachment[]
}

interface Attachment {
  '@type'?: string
  id: string
  href?: string
  description?: string
  mimeType?: string
  name?: string
  url?: string
  redirectUrl?: string
  displayInfo?: DisplayInfo
}

interface DisplayInfo {
  valueType?: string
  value?: string[]
}
